### ISSUE TITLE
Rule proposal: Require a test to have an assertion

### ISSUE THREAD

**Issue by pmcelhaney** [lo1tuma/eslint-plugin-mocha#90]
If a test has no assertions, it looks like it passed, when in fact nothing happened. It would be nice to have a rule that catches when a test does not have an assertion.

Bad

``` js
it('adds two numbers', () => {
   const numerator = 4;
   const denominator = 2;
   // oops, no assertion!
});
```

Good

``` js
it('adds two numbers', () => {
   const numerator = 4;
   const denominator = 2;
   expect(calc.divide(numerator, denominator)).to.equal(2);
});
```

By default, it could check for an identifier from one of the assertion libraries (`assert`, `expect`, `should`). Or a list of identifiers could be provided as an option.

**Comment by lo1tuma** [comment]
Hi @pmcelhaney and thanks for your request.

I’m not sure about this proposal as it would require knowledge about the assertion libraries being used. Apart from that it would be also quite hard to ensure that at least one assertion is executed in every possible code path, for example:

```js
it('async test', function (done) {
    foo(function (error, data) {
        if (error) {
            return done(); // no assertion in this branch and done was also called without error object
        }
        assert(data === 'foo');
        done();
    });
});

**Comment by jfmengels** [comment]
I would love to see this too, but it would be kind of hard.
You have the example shown by @lo1tuma, but here's another tricky one:

```js
function checkIfFoo(book) {
  assert(book);
  assert(book.type === 'book');
}

it('foo should return a book', function() {
  const book = foo();
  checkIfIsBook(book);
});

it('bar should return a book', function() {
  const book = bar();
  checkIfIsBook(book);
});
```

That said, it might be worth adding, and have users then disable the rule on sections that we can't help it. It could be pretty useful in most cases.

**Comment by pmcelhaney** [comment]
Good points. 

#### "it would require knowledge about the assertion libraries being used"

I think we can work around that issue by making that knowledge a configuration option. 

```js
'require-assertion':  ['warning', {
   assertFunctions: ['assert', 'expect', 'should']
}]
```

#### "it would be also quite hard to ensure that at least one assertion is executed in every possible code path"

I think that's okay because you still have an assertion somewhere in the test. The rule's not meant to be fool-proof. Also, I think branches within a test is an anti-pattern. Shouldn't a test be deterministic, so that it follows the same path every time?

#### What if the assertion is in a function called by the test?

That's a tough one. Maybe if a convention is established for functions that contain asserts, we could find them that way.

```js
'require-assertion':  ['warning', {
   assertFunctions: ['assert', 'expect', 'should'],
   assertContainingFunctionPattern: '^(check|assert|ensure)'
}]
```

**Comment by pmcelhaney** [comment]
Another potential problem that comes to mind is when the assertion is placed in `afterEach()`. I'm not sure how difficult it is to find all of the `afterEach`s associated with an `it`, but should be possible.

**Comment by jfmengels** [comment]
> > "it would require knowledge about the assertion libraries being used"

> I think we can work around that issue by making that knowledge a configuration option.

That would probably be a good idea. The biggest hurdle would be to add logic to detect (the most faithfully possible) whether a statement is an assertion from X library (find things looking like `xyz.should.abc` or `should(foo)` in the case of `should`).

Also, you'd have to keep up with requests to support new assertion library Y (at which point you wonder whether you should). I agree this would be a really nice rule, at least for most cases, but this would be a pain to maintain. Not saying this is a reason not to do it, but just a heads up, as I've seen this happen in other plugins.

> Maybe if a convention is established for functions that contain asserts, we could find them that way

It's a nice idea. I'd like to see how this it used in practice before this gets implemented though (getting feedback from others I mean)

> Another potential problem that comes to mind is when the assertion is placed in afterEach()

That sounds to me like a very bad practice. Assertions should be in the test, not in the clean up hook :confused:

**Comment by pmcelhaney** [comment]
> The biggest hurdle would be to add logic to detect (the most faithfully possible) whether a statement is an assertion from X library (find things looking like xyz.should.abc or should(foo) in the case of should).

I think we would simply look for the identifier "should", and not worry about whether it's a function or a property. 


> Also, you'd have to keep up with requests to support new assertion library Y (at which point you wonder whether you should).

We can avoid that by making the identifiers a schema option. (See "assertFunctions" above. Maybe not the best name, but we can work on that.)

**Comment by lo1tuma** [comment]
Closing since there haven’t been a good proposal how to achieve this in a assertion-framwork-agnostic way.
